Programクラスのパフォーマンス改善

masnagam opened the pull request:
`get`および`exists`メソッドの計算量をO(N)からO(1)に改善．

これにより，以下のパフォーマンスが改善します．

* 起動時の`program.json`の読み込み時間
* `findByNetworkIdAndReplace`のブロック時間

`add`のたびに，`_item`を線形検索していましたが，ほとんどの場合IDがマッチすることはないのでかなりの時間を消費していました．

この修正により，`findByNetworkIdAndReplace`のブロック時間は問題にならない程度に速くなったので（Rock64でも１秒以下），`findByNetworkIdAndReplace`の非同期化は不要となりました．そのため，`async`および`await`のみ削除しました．

コメントにあるように，`_items`を`_itemMap`に置き換えてしまうことも可能でしたが，そうしてしまうと，Programクラスの大部分を書き換えることになるため，置き換えはしませんでした．置き換えることで，`splice`しなくて良くなるのでもう少し速くなるかもしれませんが，すでに十分速くなっているので，必要性は低いと考えています．

`_items`と`_itemMap`という似たものが２つあるのはよくないということでしたら，`_items`を削除しますので，指摘をお願いします．

masnagam committed:
program: improve performance of the `get` and `exists` methods

Improve the time complexity of `get` and `exists` methods from O(N) to
O(1).

This change affects performance of the following functions:

* Load time of program.json
* findByNetworkIdAndReplace

New private member variable `_itemMap` has been introduced, which maps
program IDs to their associated ProgramItem objects.

masnagam committed:
program: remove useless async/await keywords

masnagam commented:
ちなみに，この修正は #56 とは無関係です（多分．．）

kanreisa commented:
対応が遅くなりました。

`_itemMap` なのですが、[Map](https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Global_Objects/Map) はいかがでしょうか？
Map 使った上で、もしパフォーマンス問題なさそうであれば、_items を消してしまってもよいと思います。

masnagam commented:
`_items`を削除するかどうかは，パフォーマンスの良し悪しで判断するのではなく，修正による影響を考慮した上で判断すべきだと考えます．

書き換えてもいいのですが，以下について問題がないか判断をお願いします．

* 最初のコメントにも書きましたが，`_items`を現在の`Array`から`Object`または`Map`に置き換えた場合，`Program.ts`を大幅に書き換えることになります
  * 恐らく，diffを確認する意味がなくなります
* `Program.all()`の動作が変わります
  * 連想配列からプログラム一覧を取得するので，毎回違うJS Objectが返されますし，要素の順番も保証されません
  * 上記のように`Program.all()`の動作が変わると，問題が発生するようなコードはないですか？

問題ないのであれば書き換えます．

`Object`も`Map`もV8ではハッシュテーブルのようなものを使って実装されているようです．つまり，要素数に対して大雑把に考えてO(1)です．そのため，どちらを使っても大して変わらないと考えています．最初のコメントにあるように，要素を追加するたびに`_items`を線形検索（O(N)）していて，ほとんど毎回全要素をサーチしていることが，速度が遅いことの原因だからです．

kanreisa commented:
まさに
> 似たものが２つあるのはよくない
と考えていますから、提案していただいたパフォーマンス改善の意図をそのまま生かす形で Map 使えたら Map 使ってみましょう

私もテストしますので、ぜひ書き換えてみてください。

kanreisa commented on src/Mirakurun/Program.ts:
```diff
@@ -47,6 +47,10 @@ export default class Program {
+    // At this moment, this is used only for improving the performance
+    // of the `get` and `exists` methods.  But it may be possible to
+    // replace _items with _itemMap and keep the API compatibility.
+    private _itemMap: { [programId: number]: ProgramItem } = {};
```
- Map のほうが良いかもしれないという件
- `_items` の配列は廃止してみましょう

masnagam replied:
* ba086a42fe88b6fbff06eb99d61770a9a93f840d で`Map`に変更しました
* 48858b37cf1cd6edfce75e22462440c1605bc80f で`_items`を削除しました

masnagam commented:
了解です．念の為，以下の順番でコミットします．

1. `Object` -> `Map`への変更のみ
2. `_items`を削除
3. `_itemMap` -> `_items`に改名

2, 3の修正がまずかったら簡単に戻せるようにしておきます．

masnagam committed:
program: change the type of `_itemMap` from `Object` to `Map`

masnagam committed:
program: remove `_items`

The `Map` class keeps insertion order.  So, we expect that the order
of program items in new `Program.items` is the same as the old one,
hopefully...

masnagam commented:
コミットログにも書きましたが，`Map`は[挿入順を記憶している](https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Global_Objects/Map/entries)みたいです．上手くいけば，`Program.items`の順番は以前と同じはず．．

masnagam commented:
`_itemMap`を`_items`に改名すると，名前が似ているものと混同してしまうので止めておきました．

masnagam committed:
add docker-compose.yml for test purpose

Dockerfile has been updated because it didn't work.  In addition, the
inconsistent line endings in Dockerfile have been fixed.

docker-compose.yml has been added for test purpose.  A data volume
which mounted at /usr/local/var/db/mirakurun is created automatically
when running `docker-compose up`.  Use `docker-compose down` with the
`-v` option if you want to remove the data volume at the shutdown.  Or
remove it manually using `docker voluem rm`.

masnagam commented:
単体テストなどがないため，とりあえず暫く動かしてみます

masnagam committed:
program: update a comment

masnagam commented:
一日程度動かしましたが，`Map`変更・`_items`削除前と比べて遅くなったりはしていないようでした．

kanreisa reviewed (changes requested):
不要な変更とファイルが混じってしまっているようです

masnagam commented:
fbc67e164bbf99118e86d8041796cd02d696158a で行ったDockerfileの変更とdocker-compose.ymlの追加でしょうか？

もしそうであれば fbc67e164bbf99118e86d8041796cd02d696158a をrevertし，別のPRとして出し直します．

masnagam committed:
Revert "add docker-compose.yml for test purpose"

This reverts commit fbc67e164bbf99118e86d8041796cd02d696158a.

masnagam commented:
1c7acea08cc448869bfe57a10fc886bfe2159cdb でrevertしました．

Dockerfileの修正とdocker-compose.ymlの追加だったので，revertしても本件の修正には影響を与えません．

masnagam commented:
分離して #59 を作成しました．

kanreisa commented:
ありがとうございます。マージしました。
(https://github.com/Chinachu/Mirakurun/blob/master/CHANGELOG.md
